File: sources/go/14_file_read.go
// task 14 file_read — expected output: 484442112
// build: go build -o prog 14_file_read.go    run: ./prog
// note: plain "go build" only; do not set GOGC=off or any other tuning flags
package main

import (
	"fmt"
	"io"
	"os"
)

func main() {
	f, err := os.Open("data.bin")
	if err != nil {
		panic(err)
	}
	defer f.Close()

	buf := make([]byte, 1<<20)
	var total uint64
	for {
		n, err := f.Read(buf)
		for i := 0; i < n; i++ {
			total += uint64(buf[i])
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			panic(err)
		}
	}
	fmt.Println(total % 4294967296)
}

File: sources/rust/14_file_read.rs
// task 14 file_read — expected output: 484442112
// build: rustc -O -o prog 14_file_read.rs    run: ./prog

use std::fs::File;
use std::io::Read;

fn main() {
    let mut file = File::open("data.bin").expect("cannot open data.bin");
    let mut buf = vec![0u8; 1024 * 1024];
    let mut total: u64 = 0;

    loop {
        let n = file.read(&mut buf).expect("read failed");
        if n == 0 {
            break;
        }
        for i in 0..n {
            total += u64::from(buf[i]);
        }
    }

    println!("{}", total % 4_294_967_296);
}
